Add sphinx-tabs + use it for OS-specific "py[thon] -m pip"

shireenrao opened the pull request:
This PR is to resolve #7311. I will change a few pages at a time and submit them to this PR. 

(Edit by @uranusjr so a merge would auto-close to issue.)

McSinyx commented on docs/html/installing.rst:
```diff
@@ -97,12 +97,7 @@ the `Python Packaging User Guide
-
-
-On Windows [4]_::
+On Windows, Linux or macOS::
```
Since this doesn't categorize anything (and implies that on other OSes the method would be different), I suggest replacing this by a filler, e.g. *in a shell* or something like that.

shireenrao replied:
I changed it to "In a shell". Thank you.

pfmoore replied:
Do we anywhere make the point that you don't always invoke Python using the `python` command? I feel as though we should clarify this *somewhere* if we're going to make `python -m pip` be the canonical way we say "run pip"...

(This is arguably off-topic for this PR, and honestly, arguably not even pip's problem to address, but I feel that without some such clarification, this change is attacking a symptom, rather than the cause, of the problem people have invoking pip).

shireenrao replied:
Adding some context on why we should use `python -m pip` should help. Should I create a new page or add it to an existing page? We can reference Brett Cannon's article the https://snarky.ca/why-you-should-use-python-m-pip/ there.

pfmoore replied:
Sorry, I wasn't clear. I'm fine with the `-m pip` bit. The problem is that on Windows, `python` probably isn't on `PATH`, you should use `py`. On Unix, you should often use `python3` rather than `python`. Unless you're in a virtual environment. Etc.

Using `-m pip` means that you don't need to think about how the `pip` command relates to whatever way you run Python. But it leaves you still needing to know how to invoke Python. We're using `python` as a shortcut for "whatever command you use to run Python" and I don't see anything explaining that (specifically "we don't necessarily mean you should type `python` literally")...

McSinyx replied:
I think there's also a need to mention that on Windows Python version is specified via `py -<major>.<minor>` instead of `python<major>.<minor>`.

shireenrao replied:
@pfmoore - Please let me know how to proceed? Does this change need further discussion before I proceed to making more changes?

pfmoore replied:
I'd like to know what other @pypa/pip-committers think.

uranusjr replied:
We can have a box on top of each page to tell the user they should use `pythonX.Y` or `py -X.Y` depending on the platform. But I don’t know how is best to annotate each code block to make it obvious to the user they should not copy-paste the command verbatim and must read the block at the top.

pradyunsg commented:
Bah, I'm not super keen on this.

I really think what we need is better capabilities in our documentation, to provide more platform specific instructions in our documentation overall without looking ugly/tacked-on (see https://squidfunk.github.io/mkdocs-material/getting-started/#with-docker-recommended for a good example).

shireenrao commented:
@pradyunsg - that will be really cool!!

shireenrao commented:
Looking around I found [sphinx-tabs](https://github.com/djungelorm/sphinx-tabs). This might be helpful to add to the documentation.

pradyunsg commented:
> Looking around I found [sphinx-tabs](https://github.com/djungelorm/sphinx-tabs). This might be helpful to add to the documentation.

Yeap -- that's the ideal thing for us to be using. I don't like the out-of-the-box styling of it TBH, so I'm gonna defer to others on the *how* of this. :)

pfmoore commented:
Agreed sphinx-tabs looks like a good way to go (if we could add some browser detection to default to the tab for the reader's OS, then even better!). And I also agree that the default style isn't ideal. But I'm no css expert so I'm not going to be the one to change it 🤷 

Also, on the importance of giving a correct command for the user's situation, we're seeing more and more reports of "`python setup.py` did nothing" on packaging-problems, which are caused by the default "python" command on Windows being the Store hook, which silently does nothing if given arguments¹. We don't want pip users hitting that issue of we can avoid it.

¹ Don't get me started on my opinion of that particular bit of behaviour 🙁

shireenrao commented:
Here is another sphinx extension which I think looks better- https://sphinxcontrib-contentui.readthedocs.io/en/latest/index.html
The [examples](https://sphinxcontrib-contentui.readthedocs.io/en/latest/tabs.html#more-examples) look promising!

shireenrao commented:
Should I close this PR and open a new issue to add tabs to the documentation?

shireenrao commented:
I added the `sphinxcontrib-contentui` extension and built the docs locally. I then changed the `Upgrading pip` section to use tabs and it looks like this:
![image](https://user-images.githubusercontent.com/887754/87885505-9eabb100-c9e4-11ea-9a82-0b68a57eceec.png)
and
![image](https://user-images.githubusercontent.com/887754/87885524-b125ea80-c9e4-11ea-90f2-896cb3a22357.png)
I uploaded the sphinx generated docs in https://github.com/shireenrao/pip-docs. Once you clone it, just go into the docs folder and then run `python -m http.server`. Then go to http://localhost:8000/installing.html#upgrading-pip to see it in action.
I was hoping that it could be visible on github pages at https://shireenrao.github.io/pip-docs/ but the site is not rendering properly.

Let me know what you guys think.

pfmoore commented:
My expectation was that the Unix instructions would be `python -m pip install -U pip` (or maybe `python3`?) and the Windows instructions would be `py -m pip install -U pip`. Is that not the point here? Also, I don't have the time to build my own copy to check, but does the display adapt to the user's OS (so that, as I use Windows, I'd see the Windows instructions by default)? Otherwise I can imagine people not realising they could click on the buttons, and assuming the default version applies everywhere.

On a more subjective note, I find the style you posted in the screenshot too distracting. Is there an alternative style that's more unobtrusive?

pradyunsg commented:
Can we stick to sphinx-tabs? I'm happy to add a tiny bit of CSS myself to make things look better.

shireenrao commented:
> My expectation was that the Unix instructions would be `python -m pip install -U pip` (or maybe `python3`?) and the Windows instructions would be `py -m pip install -U pip`.
Oops.. I was trying to showcase the tabs. Sorry for missing the real point. Detecting the User's OS is not there and I don't see any options to customize the look. The only way to change the look would be to change the css.

pradyunsg commented:
> Also, I don't have the time to build my own copy to check, but does the display adapt to the user's OS (so that, as I use Windows, I'd see the Windows instructions by default)?

We deploy the documentation for each PR: https://pip--8589.org.readthedocs.build/en/8589/

shireenrao commented:
I run the docs build using `tox -e docs`. I can see that the sphinxcontrib.contentui package does not get installed. The section in the logs showing all the packages does not show it. Eventually the build fails..

`
GLOB sdist-make: /home/srao/projects/pipdev/dev/setup.py
docs inst-nodeps: /home/srao/projects/pipdev/dev/.tox/.tmp/package/1/pip-20.2.dev1.zip
docs installed: alabaster==0.7.12,Babel==2.8.0,certifi==2020.6.20,chardet==3.0.4,docutils==0.16,idna==2.10,imagesize==1.2.0,Jinja2==2.11.2,MarkupSafe==1.1.1,packaging==20.4,pip @ file:///home/srao/projects/pipdev/dev/.tox/.tmp/package/1/pip-20.2.dev1.zip,Pygments==2.6.1,pypa-docs-theme @ git+https://github.com/pypa/pypa-docs-theme.git@d2e63fbfc62af3b7050f619b2f5bb8658985b931,pyparsing==2.4.7,python-docs-theme @ git+https://github.com/python/python-docs-theme.git@628a66466589ba42037d4aad3d676cb7d81fb6e6,pytz==2020.1,requests==2.24.0,setuptools==49.2.0,six==1.15.0,snowballstemmer==2.0.0,Sphinx==2.4.3,sphinxcontrib-applehelp==1.0.2,sphinxcontrib-devhelp==1.0.2,sphinxcontrib-htmlhelp==1.0.3,sphinxcontrib-jsmath==1.0.1,sphinxcontrib-qthelp==1.0.3,sphinxcontrib-serializinghtml==1.1.4,urllib3==1.25.9,wheel==0.34.2
docs run-test-pre: PYTHONHASHSEED='3354392635'
......
Extension error:
Could not import extension sphinxcontrib.contentui (exception: No module named 'sphinxcontrib.contentui')
ERROR: InvocationError for command /home/srao/projects/pipdev/dev/.tox/docs/bin/sphinx-build -W -d /home/srao/projects/pipdev/dev/.tox/docs/tmp/doctrees/html -b html docs/html docs/build/html (exited with code 2)
_______________________________________________________ summary ________________________________________________________
ERROR:   docs: commands failed`

Thats why I havn't pushed this change yet. I have edited docs/html/conf.py to add the sphinxcontrib.contentui extension in the extensions list a the end. I also added this as a docs build requirement under tools/requirements/docs.txt. What else am I missing?

shireenrao commented:
> Can we stick to sphinx-tabs? I'm happy to add a tiny bit of CSS myself to make things look better.

Sure. I will try and add it.

pradyunsg commented:
If you're changing dependencies for the tox job, you'll have to pass -r as well, to tell it to recreate the environment.

shireenrao commented:
> If you're changing dependencies for the tox job, you'll have to pass -r as well, to tell it to recreate the environment.

Thats it!! That Solved the issue! 🙏

shireenrao commented:
I am not sure what is causing the man pages build to fail. The html pages get built fine. Here is the log for man pages section -

docs run-test: commands[1] | sphinx-build -W -d /home/srao/projects/pipdev/pip/.tox/docs/tmp/doctrees/man -b man docs/man docs/build/man -c docs/html
Running Sphinx v2.4.3
20.2
loading intersphinx inventory from https://packaging.python.org/objects.inv...
loading intersphinx inventory from https://www.pypa.io/en/latest/objects.inv...
building [mo]: targets for 0 po files that are out of date
building [man]: all manpages
updating environment: [new config] 15 added, 0 changed, 0 removed
reading sources... [  6%] commands/cache
reading sources... [ 13%] commands/check
reading sources... [ 20%] commands/config
reading sources... [ 26%] commands/debug
reading sources... [ 33%] commands/download
reading sources... [ 40%] commands/freeze
reading sources... [ 46%] commands/hash
reading sources... [ 53%] commands/help
reading sources... [ 60%] commands/install
reading sources... [ 66%] commands/list
reading sources... [ 73%] commands/search
reading sources... [ 80%] commands/show
reading sources... [ 86%] commands/uninstall
reading sources... [ 93%] commands/wheel
reading sources... [100%] index

looking for now-outdated files... none found
pickling environment... done
checking consistency... done
writing... pip.1 { } pip-help.1 { } pip-show.1 { } pip-freeze.1 { } pip-wheel.1 { } pip-debug.1 { } pip-cache.1 { } pip-config.1 { } pip-list.1 { } pip-check.1 { } pip-download.1 { } pip-uninstall.1 { } pip-search.1 { } pip-install.1 { } pip-hash.1 { } done
build succeeded.

The manual pages are in docs/build/man.

Warning, treated as error:
Not copying tabs assets! Not compatible with man builder
ERROR: InvocationError for command /home/srao/projects/pipdev/pip/.tox/docs/bin/sphinx-build -W -d /home/srao/projects/pipdev/pip/.tox/docs/tmp/doctrees/man -b man docs/man docs/build/man -c docs/html (exited with code 2)
_______________________________________________________ summary ________________________________________________________
ERROR:   docs: commands failed

Any thoughts?

shireenrao commented:
I found the issue in the sphinx-tabs extension. I have requested a PR there to fix this issue. Man pages are not setup to be a build target in that project and is setup to throw a warning when a new target is used.  Currently they have the following targets - 
['html', 'singlehtml', 'dirhtml', 'readthedocs', 'readthedocsdirhtml', 'readthedocssinglehtml', 'readthedocssinglehtmllocalmedia', 'spelling']. I have requested `man` be added to this list.

xavfernandez commented on tools/requirements/docs.txt:
```diff
@@ -1,6 +1,7 @@
 sphinx == 2.4.3
 git+https://github.com/python/python-docs-theme.git#egg=python-docs-theme
 git+https://github.com/pypa/pypa-docs-theme.git#egg=pypa-docs-theme
+https://github.com/djungelorm/sphinx-tabs/releases/download/v1.1.13/sphinx-tabs-1.1.13.tar.gz
```
why not `sphinx-tabs==1.1.13` ?

shireenrao replied:
fixed.

shireenrao commented:
@pradyunsg - is it ok for the sake of moving forward that I forked the sphinx-tabs project and make the necessary changes and reference my forked version in this PR temporarily? This will also let me play with the css/js to fine tune the look of the tabs.

pradyunsg commented:
> is it ok for the sake of moving forward that I forked the sphinx-tabs project and make the necessary changes and reference my forked version in this PR temporarily?

I'm not sure I follow -- what issue are you working around?

To answer your question more directly, no, I don't it'd make sense to have a fork to this. It's possible to add/override a CSS file with `html_css_files` in Sphinx (and similar for js) so I'm not sure we need a fork to tweak the JS/CSS behaviors.

shireenrao commented:
> I have requested `man` be added to this list with this [PR](https://github.com/djungelorm/sphinx-tabs/pull/72).

As suggest in the sphinx-tabs PR, I have closed it. Looking at the history of the sphinx-tabs project, the warning is thrown on purpose to only support platforms the tabs work on. 

I am not sure if one of the two is possible 

- have two different sphinx conf.py files one for html build and one for man pages build
- have conf.py be aware of what kind of build is being executed and enable the appropriate extension based on type of build

shireenrao commented:
> To answer your question more directly, no, I don't it'd make sense to have a fork to this. It's possible to add/override a CSS file with `html_css_files` in Sphinx (and similar for js) so I'm not sure we need a fork to tweak the JS/CSS behaviors.

Cool. I wasn't aware of it.

shireenrao commented:
Thank you for being patient with my changes as I am still learning the ropes. If what I did by creating a new conf.py for man pages is acceptable, then I can trim the config files even more to be specific to each type of build. Please let me know if I am going in the right direction.

shireenrao commented:
I was reading the documentation for sphinx-build and saw that I can override settings using the -D option. so I will undo the last changes I made and just resubmit with just the tox.ini changes to use the -D for man build.

shireenrao commented:
Here are some designs I came up with -
1. 
![image](https://user-images.githubusercontent.com/887754/88096155-3552ac00-cb64-11ea-9088-e5d8bd4836b5.png)
2. 
![image](https://user-images.githubusercontent.com/887754/88096267-603d0000-cb64-11ea-8b27-6c20245c3975.png)
3. Default - 
![image](https://user-images.githubusercontent.com/887754/88096372-8793cd00-cb64-11ea-8dce-66d8be07dea2.png)

Any preferences?

pfmoore commented:
Personally I prefer (3).

shireenrao commented:
Hello @pradyunsg  and @pfmoore - How should I proceed with this PR?

uranusjr commented on docs/html/quickstart.rst:
```diff
@@ -8,7 +8,7 @@ Install a package from `PyPI`_:
 ::
 
-  $ pip install SomePackage
+  $ python -m pip install SomePackage
```
Shouldn’t these also be updated to use tabs?

uranusjr commented:
It seems like people are quite satisfied by the tab solution (option 3). So I’d say let’s update the PR to change all the `python -m pip` blocks to use tabs, and this should be ready.

shireenrao commented:
Should the Linux/MacOS tab be called Unix instead?

shireenrao commented:
I started updating the quickstart document and adding the multiline texts within a tab is going to look like this 
![image](https://user-images.githubusercontent.com/887754/88698403-01333a00-d0d4-11ea-824f-dbf0707fcabe.png)

Is this still better compared to this:
![image](https://user-images.githubusercontent.com/887754/88698167-b3b6cd00-d0d3-11ea-8476-b4594b69ab2b.png)

uranusjr commented:
Unix sounds fine, but I would keep the Mac part in the tab (e.g. Unix/Mac) for less experienced users. The bottom image (tab + code block) looks good to me.

pradyunsg commented:
> adding the multiline texts within a tab is going to look like this

FWIW, I think they're all not-great-looking, but the latter (code block inside tab) is definitely more appropriate (since it's a monospaced text for stuff from the terminal).

I say we ship the latter and worry about making our documentation prettier in a follow up "mega-esque" task that I have a feeling I will pick up. :)

pradyunsg commented on tox.ini:
```diff
@@ -56,7 +56,7 @@ commands =
     # currently -- https://github.com/rtfd/readthedocs.org/issues/1543.
     # That is why we have a "-c docs/html" in the next line.
-    sphinx-build -W -d {envtmpdir}/doctrees/man -b man docs/man docs/build/man -c docs/html
+    sphinx-build -W -d {envtmpdir}/doctrees/man -b man docs/man docs/build/man -c docs/html -D extensions=sphinx.ext.extlinks,pip_sphinxext,sphinx.ext.intersphinx
```
Oh no. :/

Let's merge this in for now, and I'll clean up the mess that is our conf.py situation in a follow up PR. :)

pradyunsg commented on news/7311.doc:
```diff
@@ -0,0 +1 @@
+Use ``python -m`` to run pip in the docs
```
I'd make this a .trivial file TBH. Or change it to:

```
Add tabs to the documentation UI.
```

shireenrao commented:
In the [Running pip](https://pip.pypa.io/en/stable/user_guide/#running-pip) section of the user-guide, does it make more sense to change the current 3 code blocks to a single tabbed code block and remove this paragraph

> Even though pip is available from your Python installation as an importable module, via import pip, it is not supported to use pip in this way. For more details, see Fixing conflicting dependencies.

Instead have a few lines, similar to:

>`python -m pip` on Unix/Mac or `py -m pip` on windows, executes pip using the 
Python interpreter you specified as python. So `/usr/bin/python3.7 -m pip` means
you are executing pip for your interpreter located at /usr/bin/python3.7. 

>On Windows it is recommended to invoke pip using the Python windows launcher `py`.
For more details, see https://docs.python.org/3/using/windows.html#launcher

McSinyx commented:
I suggest having the

> `python -m pip` on Unix/Mac executes pip using the
> Python interpreter you specified as `python`. So `/usr/bin/python3.7 -m pip` means
> you are executing pip for your interpreter located at `/usr/bin/python3.7`.

part in the Unix/Mac tab and the equivalent for Windows.  Bike shedding a little: the Apple's OS goes by macOS nowadays, and Unix does not cover the most popular consumer-grade OSes like GNU/Linux and FreeBSD, but for brevity it might be best for this situation.

shireenrao commented:
@McSinyx - Please take a look at https://pip--8589.org.readthedocs.build/en/8589/user_guide/#running-pip

Is this ok? I also change wherever I had put the tab to be Mac as macOS.

McSinyx commented:
Personally I'd prefer the paragraph not being formatted as code if that is possible.

shireenrao commented:
Unfortunately it's not possible without customizing the css. If this is looking too ugly, I can move the two paragraphs below the code tab layout.

McSinyx commented:
@shireenrao, I believe that it is possible to have `code-block` inside a `tab` (or maybe preferably `group-tab`?).

shireenrao commented:
I am now using a code-block within a group-tag. I made that change everywhere to make the look consistent.

shireenrao commented:
The only docs left to do are the reference docs. These docs are using custom directives. For example in the Usage section for [pip install](https://pip.pypa.io/en/stable/reference/pip_install/#usage), that entire code block is a custom directive which comes from executing create_command by passing "install". create_command is coming from `from pip._internal.commands import`. Is it ok to add two different directives for each command? One for windows and one for unix/macOS? For instance, right now to print the pip usage for install, we have the custom directive called `pip-command-usage`. I am proposing to change that to be two directives

     pip-windows-command-usage - This will print `py -m pip` instead of just `pip` for all command usages blocks
     pip-unix-command-usage - This will print `python -m pip`

I would really appreciate guidance on how to proceed.

pradyunsg commented on docs/html/user_guide.rst:
```diff
@@ -358,15 +499,15 @@ all users) configuration:
 **Inside a virtualenv**:
 
-* On Unix and macOS the file is :file:`$VIRTUAL_ENV/pip.conf`
+* On Unix and macOSOS the file is :file:`$VIRTUAL_ENV/pip.conf`
```
Find-replace gone wrong?

shireenrao replied:
Oops! 🙈 
Fixed now.

uranusjr commented:
Or we can modify the directive toe accept an optional flag, so we can write

```rst
.. pip-command-usage:: install
   :pip: py -m pip

.. pip-command-usage:: install
   :pip: python -m pip
```

shireenrao commented:
@uranusjr - please take a look at how I implemented the directive using optional flags in the [pip install reference docs](https://pip--8589.org.readthedocs.build/en/8589/reference/pip_install/)

uranusjr commented on docs/pip_sphinxext.py:
```diff
@@ -15,11 +15,15 @@
         cmd = create_command(self.arguments[0])
+        pip_cmd = '$ python -m pip'
+        if len(self.arguments) > 1:
+            pip_cmd = " ".join(self.arguments[1:])
```
I feel it’d be better if we do

```rst
.. pip-command-usage:: install "python -m pip"
```

instead. The current argument format feels confusing since there’s no logic separation between the pip command (`install`) and the CLI prefix.

Also, I feel the `$` and `C:\>` prefix unnecessary here.

shireenrao replied:
I was not a fan as well. I added the quotes as suggested.

shireenrao commented:
@uranusjr - all docs have been updated. I am still going through all the docs to see if I missed any.. please take a look when you get a chance.

uranusjr commented:
The `python get-pip.py` calls may needed to be updated to use tabs as well. Otherwise this looks great, thanks so much for the effort! I love how the tab switch is synced across all code blocks (which I assume is a feature from the extension?)

shireenrao commented:
> The `python get-pip.py` calls may needed to be updated to use tabs as well. Otherwise this looks great, thanks so much for the effort! I love how the tab switch is synced across all code blocks (which I assume is a feature from the extension?)

The tab switch across the page works because of the use of the group-tab construct in sphinx-tabs. I updated the `python get-pip.py` section as well. Glad this was helpful.

pradyunsg commented on tox.ini:
```diff
@@ -56,7 +56,7 @@ commands =
     # currently -- https://github.com/rtfd/readthedocs.org/issues/1543.
     # That is why we have a "-c docs/html" in the next line.
-    sphinx-build -W -d {envtmpdir}/doctrees/man -b man docs/man docs/build/man -c docs/html
+    sphinx-build -W -d {envtmpdir}/doctrees/man -b man docs/man docs/build/man -c docs/html -D extensions=sphinx.ext.extlinks,pip_sphinxext,sphinx.ext.intersphinx
```
We should set `sphinx_tabs_nowarn = True` in conf.py, instead of modifying this.

shireenrao replied:
I have removed the changes to tox.ini and updated conf.py to add `sphinx_tabs_nowarn = True`

hugovk commented on docs/html/quickstart.rst:
```diff
@@ -6,62 +6,145 @@ First, :doc:`install pip <installing>`.
 
 Install a package that's already been downloaded from `PyPI`_ or
-obtained from elsewhere. This is useful if the target machine does not have a
+obtained from elsewhere. This is useful if the target macOShine does not have a
```
Whoops!
```suggestion
obtained from elsewhere. This is useful if the target machine does not have a
```

hugovk replied:
Please change `macOShine` back to `machine`:

```suggestion
obtained from elsewhere. This is useful if the target machine does not have a
```

hugovk commented on docs/html/development/getting-started.rst:
```diff
@@ -38,13 +38,25 @@ To run the pip executable from your source tree during development, install pip
+      .. code-block:: shell
 
+         C:\> virtualenv venv # You can also use "py -m venv venv" from python3.3+
+         C:\> source venv/bin/activate
```
Does `source` work on Windows?

https://stackoverflow.com/q/8921188/724176

shireenrao replied:
Thank you. it should be `venv\Scripts\activate` instead. Will push this change shortly.

hugovk commented:
A suggestion:

* if example commands have output, use prompts (and `.. code-block:: console`)
* otherwise omit prompts, so the user can copy and paste the whole line (or lines) without needing to delete anything (and use `.. code-block:: shell`)

The `console` or `shell` code-block will help colour the example, for readability.

A couple of examples. This one (from getting-started.html) currently has prompts and no output. Remove the prompts:

| Unix/macOS | Windows |
| - | - |
| ![image](https://user-images.githubusercontent.com/1324225/90222123-16b69c80-de14-11ea-9580-825ab11bd66c.png) | ![image](https://user-images.githubusercontent.com/1324225/90222159-2c2bc680-de14-11ea-8c61-da64e4186790.png) |

file:///Users/hugo/github/pip/docs/build/html/development/getting-started.html


This (from quickstart.html) has prompts and output. Instead of `code-block:: shell`, use `code-block:: console`:

| Unix/macOS | Windows |
| - | - |
| ![image](https://user-images.githubusercontent.com/1324225/90222362-97759880-de14-11ea-8615-b2c94e25db91.png) | ![image](https://user-images.githubusercontent.com/1324225/90222380-9d6b7980-de14-11ea-8d0e-8214d78b7db5.png) |

shireenrao commented:
> This (from quickstart.html) has prompts and output. Instead of `code-block:: shell`, use `code-block:: console`:

interestingly on the windows tab having `code-block:: shell` has better color and having `code-block:: console` on the unix tab has better color. Any thoughts on that @hugovk?

shireenrao commented:
@hugovk - I pushed changes so that only commands which have outputs have a platform prompt. I removed it from everywhere else I had added. The only thing left to resolve is the question about the type of code-block to use. Does it make sense to use `console` for unix tabs and `shell` for windows tabs?

hugovk commented:
Hmm, I guess there's not a specific one for Windows?

I think `console` is better for Windows prompts, with `shell`, it's a bit strange that `:>` and `1` and `for` are (mis)coloured here:


![image](https://user-images.githubusercontent.com/1324225/90341984-1a614380-e00d-11ea-9f58-dd2d029390ac.png)

hugovk commented on docs/html/quickstart.rst:
```diff
@@ -6,62 +6,145 @@ First, :doc:`install pip <installing>`.
+
+   .. group-tab:: Unix/macOS
+
+      .. code-block:: shell
```
```suggestion
      .. code-block:: console
```

hugovk commented on docs/html/quickstart.rst:
```diff
@@ -6,62 +6,145 @@ First, :doc:`install pip <installing>`.
+
+   .. group-tab:: Windows
+
+      .. code-block:: shell
```
```suggestion
      .. code-block:: console
```

hugovk commented on docs/html/quickstart.rst:
```diff
@@ -6,62 +6,145 @@ First, :doc:`install pip <installing>`.
+
+   .. group-tab:: Unix/macOS
+
+      .. code-block:: shell
```
```suggestion
      .. code-block:: console
```

hugovk commented on docs/html/quickstart.rst:
```diff
@@ -6,62 +6,145 @@ First, :doc:`install pip <installing>`.
-  Files:
-    ../somepackage/__init__.py
-    [...]
+      .. code-block:: shell
```
```suggestion
      .. code-block:: console
```

hugovk commented on docs/html/quickstart.rst:
```diff
@@ -6,62 +6,145 @@ First, :doc:`install pip <installing>`.
+
+   .. group-tab:: Unix/macOS
+
+      .. code-block:: shell
```
```suggestion
      .. code-block:: console
```

hugovk commented on docs/html/quickstart.rst:
```diff
@@ -6,62 +6,145 @@ First, :doc:`install pip <installing>`.
-  SomePackage (Current: 1.0 Latest: 2.0)
+   .. group-tab:: Windows
+
+      .. code-block:: shell
```
```suggestion
      .. code-block:: console
```

hugovk commented on docs/html/quickstart.rst:
```diff
@@ -6,62 +6,145 @@ First, :doc:`install pip <installing>`.
+
+   .. group-tab:: Unix/macOS
+
+      .. code-block:: shell
```
```suggestion
      .. code-block:: console
```

hugovk commented on docs/html/quickstart.rst:
```diff
@@ -6,62 +6,145 @@ First, :doc:`install pip <installing>`.
+
+   .. group-tab:: Windows
+
+      .. code-block:: shell
```
```suggestion
      .. code-block:: console
```

hugovk commented on docs/html/quickstart.rst:
```diff
@@ -6,62 +6,145 @@ First, :doc:`install pip <installing>`.
+
+   .. group-tab:: Unix/macOS
+
+      .. code-block:: shell
```
```suggestion
      .. code-block:: console
```

hugovk commented on docs/html/quickstart.rst:
```diff
@@ -6,62 +6,145 @@ First, :doc:`install pip <installing>`.
+
+   .. group-tab:: Windows
+
+      .. code-block:: shell
```
```suggestion
      .. code-block:: console
```

hugovk commented on docs/html/reference/pip_check.rst:
```diff
@@ -24,27 +32,66 @@ Examples
-     No broken requirements found.
-     $ echo $?
-     0
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_check.rst:
```diff
@@ -24,27 +32,66 @@ Examples
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_check.rst:
```diff
@@ -24,27 +32,66 @@ Examples
-     pyramid 1.5.2 requires WebOb, which is not installed.
-     $ echo $?
-     1
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_check.rst:
```diff
@@ -24,27 +32,66 @@ Examples
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_check.rst:
```diff
@@ -24,27 +32,66 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_check.rst:
```diff
@@ -24,27 +32,66 @@ Examples
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_download.rst:
```diff
@@ -69,51 +90,118 @@ Examples
-      pip-8.1.1-py2.py3-none-any.whl
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_download.rst:
```diff
@@ -69,51 +90,118 @@ Examples
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
-     jedi    0.9.0
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
-     setuptools 20.6.7  21.0.0 wheel
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
 
-     $ pip list --outdated --not-required
-     docutils (Current: 0.10 Latest: 0.11)
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
-     docopt (0.6.2)
-     idlex (1.13)
-     jedi (0.9.0)
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_uninstall.rst:
```diff
@@ -30,11 +38,26 @@ Examples
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_uninstall.rst:
```diff
@@ -30,11 +38,26 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_show.rst:
```diff
@@ -30,58 +38,124 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_show.rst:
```diff
@@ -30,58 +38,124 @@ Examples
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_show.rst:
```diff
@@ -30,58 +38,124 @@ Examples
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_show.rst:
```diff
@@ -30,58 +38,124 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_search.rst:
```diff
@@ -30,8 +38,20 @@ Examples
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_search.rst:
```diff
@@ -30,8 +38,20 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_hash.rst:
```diff
@@ -39,13 +47,32 @@ Options
+
+  .. group-tab:: Windows
+
+    .. code-block:: shell
```
```suggestion
    .. code-block:: console
```

hugovk commented on docs/html/reference/pip_hash.rst:
```diff
@@ -39,13 +47,32 @@ Options
+
+  .. group-tab:: Unix/macOS
+
+    .. code-block:: shell
```
```suggestion
    .. code-block:: console
```

hugovk commented on docs/html/reference/pip_freeze.rst:
```diff
@@ -31,19 +39,45 @@ Examples
-     Sphinx==1.2.2
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

hugovk commented on docs/html/reference/pip_freeze.rst:
```diff
@@ -31,19 +39,45 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
        .. code-block:: console
```

shireenrao commented:
@hugovk - all suggested changes have been made.

hugovk reviewed (approved):
Thanks!

shireenrao commented:
Pulled in latest changes from Upstream to make sure build was clean and using latest version.

McSinyx commented on docs/html/reference/pip_download.rst:
```diff
@@ -69,51 +90,118 @@ Examples
+      .. group-tab:: Unix/macOS
 
-      $ pip download \
+        .. code-block:: shell
+
+          python -m pip download \
           --only-binary=:all: \
           --platform any \
           --python-version 3 \
           --implementation py \
           --abi none \
           SomePackage
```
```suggestion
         .. code-block:: shell

            python -m pip download \
              --only-binary=:all: \
              --platform any \
              --python-version 3 \
              --implementation py \
              --abi none \
              SomePackage
```

McSinyx commented on docs/html/reference/pip_download.rst:
```diff
@@ -69,51 +90,118 @@ Examples
 
+      .. group-tab:: Windows
+
+        .. code-block:: shell
+
+          py -m pip download ^
+          --only-binary=:all: ^
+          --platform any ^
+          --python-version 3 ^
+          --implementation py ^
+          --abi none ^
+          SomePackage
```
```suggestion
         .. code-block:: shell

            py -m pip download ^
              --only-binary=:all: ^
              --platform any ^
              --python-version 3 ^
              --implementation py ^
              --abi none ^
              SomePackage
```

McSinyx commented on docs/html/reference/pip_download.rst:
```diff
@@ -69,51 +90,118 @@ Examples
 #. Even when overconstrained, this will still correctly fetch the pip universal wheel.
 
-    ::
+    .. tabs::
```
```suggestion
   .. tabs::
```

McSinyx commented on docs/html/reference/pip_download.rst:
```diff
@@ -69,51 +90,118 @@ Examples
+            --platform linux_x86_64 \
+            --python-version 33 \
+            --implementation cp \
+            --abi cp34m \
+            pip>=8
+
+          $ ls pip-8.1.1-py2.py3-none-any.whl
+          pip-8.1.1-py2.py3-none-any.whl
```
I'm not happy about this, but don't have any suggestion other than giving `ls` another code-block.

McSinyx commented on docs/html/reference/pip_freeze.rst:
```diff
@@ -31,19 +39,45 @@ Examples
 #. Generate output suitable for a requirements file.
 
-    ::
+    .. tabs::
```
```suggestion
   .. tabs::
```

McSinyx commented on docs/html/reference/pip_freeze.rst:
```diff
@@ -31,19 +39,45 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: console
```
```suggestion
         .. code-block:: console
```

McSinyx commented on docs/html/reference/pip_freeze.rst:
```diff
@@ -31,19 +39,45 @@ Examples
-     Sphinx==1.2.2
+      .. group-tab:: Windows
+
+        .. code-block:: console
```
```suggestion
         .. code-block:: console
```

McSinyx commented on docs/html/reference/pip_freeze.rst:
```diff
@@ -31,19 +39,45 @@ Examples
 #. Generate a requirements file and then install from it in another environment.
 
-    ::
+    .. tabs::
```
```suggestion
   .. tabs::
```

McSinyx commented on docs/html/reference/pip_freeze.rst:
```diff
@@ -31,19 +39,45 @@ Examples
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
         .. code-block:: shell
```

McSinyx commented on docs/html/reference/pip_freeze.rst:
```diff
@@ -31,19 +39,45 @@ Examples
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
         .. code-block:: shell
```

McSinyx commented on docs/html/reference/pip_hash.rst:
```diff
@@ -39,13 +47,32 @@ Options
+
+.. tabs::
+
+  .. group-tab:: Unix/macOS
+
+    .. code-block:: console
+
+      $ python -m pip download SomePackage
+          Collecting SomePackage
+            Downloading SomePackage-2.2.tar.gz
+            Saved ./pip_downloads/SomePackage-2.2.tar.gz
+          Successfully downloaded SomePackage
+      $ python -m pip hash ./pip_downloads/SomePackage-2.2.tar.gz
+          ./pip_downloads/SomePackage-2.2.tar.gz:
+          --hash=sha256:93e62e05c7ad3da1a233def6731e8285156701e3419a5fe279017c429ec67ce0
+
+  .. group-tab:: Windows
+
+    .. code-block:: console
```
```suggestion
   .. group-tab:: Unix/macOS

      .. code-block:: console

         $ python -m pip download SomePackage
         Collecting SomePackage
           Downloading SomePackage-2.2.tar.gz
           Saved ./pip_downloads/SomePackage-2.2.tar.gz
         Successfully downloaded SomePackage
         $ python -m pip hash ./pip_downloads/SomePackage-2.2.tar.gz
         ./pip_downloads/SomePackage-2.2.tar.gz:
         --hash=sha256:93e62e05c7ad3da1a233def6731e8285156701e3419a5fe279017c429ec67ce0

   .. group-tab:: Windows

      .. code-block:: console
```

McSinyx commented on docs/html/reference/pip_hash.rst:
```diff
@@ -39,13 +47,32 @@ Options
-    $ pip hash ./pip_downloads/SomePackage-2.2.tar.gz
-        ./pip_downloads/SomePackage-2.2.tar.gz:
-        --hash=sha256:93e62e05c7ad3da1a233def6731e8285156701e3419a5fe279017c429ec67ce0
+      C:\> py -m pip download SomePackage
+          Collecting SomePackage
+            Downloading SomePackage-2.2.tar.gz
+            Saved ./pip_downloads/SomePackage-2.2.tar.gz
+          Successfully downloaded SomePackage
+      C:\> py -m pip hash ./pip_downloads/SomePackage-2.2.tar.gz
+          ./pip_downloads/SomePackage-2.2.tar.gz:
+          --hash=sha256:93e62e05c7ad3da1a233def6731e8285156701e3419a5fe279017c429ec67ce0
```
```suggestion
         C:\> py -m pip download SomePackage
         Collecting SomePackage
           Downloading SomePackage-2.2.tar.gz
           Saved ./pip_downloads/SomePackage-2.2.tar.gz
         Successfully downloaded SomePackage
         C:\> py -m pip hash ./pip_downloads/SomePackage-2.2.tar.gz
         ./pip_downloads/SomePackage-2.2.tar.gz:
         --hash=sha256:93e62e05c7ad3da1a233def6731e8285156701e3419a5fe279017c429ec67ce0
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -89,15 +98,33 @@ implementation (which might possibly change later) has it such that the first
+
+      .. code-block:: console
+
+        $ python -m pip install quux
+        ...
+        Installing collected packages baz, bar, foo, quux
```
```suggestion
         $ python -m pip install quux
         ...
         Installing collected packages baz, bar, foo, quux
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -89,15 +98,33 @@ implementation (which might possibly change later) has it such that the first
-    pip install quux
-    ...
-    Installing collected packages baz, bar, foo, quux
+        $ python -m pip install bar
+        ...
+        Installing collected packages foo, baz, bar
```
```suggestion
         $ python -m pip install bar
         ...
         Installing collected packages foo, baz, bar
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -89,15 +98,33 @@ implementation (which might possibly change later) has it such that the first
+
+      .. code-block:: console
+
+        C:\> py -m pip install quux
+        ...
+        Installing collected packages baz, bar, foo, quux
+
+        C:\> py -m pip install bar
+        ...
+        Installing collected packages foo, baz, bar
```
```suggestion
         C:\> py -m pip install quux
         ...
         Installing collected packages baz, bar, foo, quux

         C:\> py -m pip install bar
         ...
         Installing collected packages foo, baz, bar
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -387,9 +414,21 @@ If your repository layout is::
-   $ pip install -e "vcs+protocol://repo_url/#egg=pkg&subdirectory=pkg_dir"
+      .. code-block:: shell
+
+          python -m pip install -e "vcs+protocol://repo_url/#egg=pkg&subdirectory=pkg_dir"
```
```suggestion
         python -m pip install -e "vcs+protocol://repo_url/#egg=pkg&subdirectory=pkg_dir"
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -387,9 +414,21 @@ If your repository layout is::
+
+      .. code-block:: shell
+
+          py -m pip install -e "vcs+protocol://repo_url/#egg=pkg&subdirectory=pkg_dir"
```
```suggestion
         py -m pip install -e "vcs+protocol://repo_url/#egg=pkg&subdirectory=pkg_dir"
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -641,17 +680,38 @@ against any requirement not only checks that hash but also activates a global
+
+      .. code-block:: console
+
+          $ python -m pip install --require-hashes -r requirements.txt
+              ...
+              Hashes are required in --require-hashes mode (implicitly on when a hash is
+              specified for any package). These requirements were missing hashes,
+              leaving them open to tampering. These are the hashes the downloaded
+              archives actually had. You can add lines like these to your requirements
+              files to prevent tampering.
+                  pyelasticsearch==1.0 --hash=sha256:44ddfb1225054d7d6b1d02e9338e7d4809be94edbe9929a2ec0807d38df993fa
+                  more-itertools==2.2 --hash=sha256:93e62e05c7ad3da1a233def6731e8285156701e3419a5fe279017c429ec67ce0
```
```suggestion
         $ python -m pip install --require-hashes -r requirements.txt
         ...
         Hashes are required in --require-hashes mode (implicitly on when a hash is
         specified for any package). These requirements were missing hashes,
         leaving them open to tampering. These are the hashes the downloaded
         archives actually had. You can add lines like these to your requirements
         files to prevent tampering.
           pyelasticsearch==1.0 --hash=sha256:44ddfb1225054d7d6b1d02e9338e7d4809be94edbe9929a2ec0807d38df993fa
           more-itertools==2.2 --hash=sha256:93e62e05c7ad3da1a233def6731e8285156701e3419a5fe279017c429ec67ce0
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -641,17 +680,38 @@ against any requirement not only checks that hash but also activates a global
+
+      .. code-block:: console
+
+          C:\> py -m pip install --require-hashes -r requirements.txt
+              ...
+              Hashes are required in --require-hashes mode (implicitly on when a hash is
+              specified for any package). These requirements were missing hashes,
+              leaving them open to tampering. These are the hashes the downloaded
+              archives actually had. You can add lines like these to your requirements
+              files to prevent tampering.
+                  pyelasticsearch==1.0 --hash=sha256:44ddfb1225054d7d6b1d02e9338e7d4809be94edbe9929a2ec0807d38df993fa
+                  more-itertools==2.2 --hash=sha256:93e62e05c7ad3da1a233def6731e8285156701e3419a5fe279017c429ec67ce0
```
```suggestion
         C:\> py -m pip install --require-hashes -r requirements.txt
         ...
         Hashes are required in --require-hashes mode (implicitly on when a hash is
         specified for any package). These requirements were missing hashes,
         leaving them open to tampering. These are the hashes the downloaded
         archives actually had. You can add lines like these to your requirements
         files to prevent tampering.
           pyelasticsearch==1.0 --hash=sha256:44ddfb1225054d7d6b1d02e9338e7d4809be94edbe9929a2ec0807d38df993fa
           more-itertools==2.2 --hash=sha256:93e62e05c7ad3da1a233def6731e8285156701e3419a5fe279017c429ec67ce0
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -697,14 +757,38 @@ Hash-checking mode also works with :ref:`pip download` and :ref:`pip wheel`. A
+
+    .. tabs::
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
+
+           python -m pip install --no-deps -e .
+
+      .. group-tab:: Windows
```
```suggestion
       .. group-tab:: Unix/macOS

          .. code-block:: shell

             python -m pip install --no-deps -e .

       .. group-tab:: Windows
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -697,14 +757,38 @@ Hash-checking mode also works with :ref:`pip download` and :ref:`pip wheel`. A
+      .. group-tab:: Windows
 
-        pip install --no-deps -e .
+        .. code-block:: shell
```
```suggestion
          .. code-block:: shell
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -697,14 +757,38 @@ Hash-checking mode also works with :ref:`pip download` and :ref:`pip wheel`. A
+        .. code-block:: shell
 
-    Instead of ``python setup.py install``, use... ::
+           py -m pip install --no-deps -e .
```
```suggestion
             py -m pip install --no-deps -e .
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -697,14 +757,38 @@ Hash-checking mode also works with :ref:`pip download` and :ref:`pip wheel`. A
+
+    .. tabs::
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
+
+           python -m pip install --no-deps .
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
+
+           py -m pip install --no-deps .
```
```suggestion
       .. group-tab:: Unix/macOS

          .. code-block:: shell

             python -m pip install --no-deps .

       .. group-tab:: Windows

          .. code-block:: shell

             py -m pip install --no-deps .
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -741,10 +838,24 @@ being copied.
+
+      .. code-block:: shell
+
+        python -m pip install -e path/to/SomeProject
+        python -m pip install -e git+http://repo/my_project.git#egg=SomeProject
```
```suggestion
         python -m pip install -e path/to/SomeProject
         python -m pip install -e git+http://repo/my_project.git#egg=SomeProject
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -741,10 +838,24 @@ being copied.
+
+      .. code-block:: shell
+
+        py -m pip install -e path/to/SomeProject
+        py -m pip install -e git+http://repo/my_project.git#egg=SomeProject
```
```suggestion
         py -m pip install -e path/to/SomeProject
         py -m pip install -e git+http://repo/my_project.git#egg=SomeProject
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -851,113 +962,292 @@ Examples
 #. Install ``SomePackage`` and its dependencies from `PyPI`_ using :ref:`Requirement Specifiers`
 
-    ::
+    .. tabs::
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: shell
```
```suggestion
   .. tabs::

      .. group-tab:: Unix/macOS

         .. code-block:: shell
```

McSinyx commented on docs/html/reference/pip_install.rst:
```diff
@@ -851,113 +962,292 @@ Examples
-      $ pip install SomePackage            # latest version
-      $ pip install SomePackage==1.0.4     # specific version
-      $ pip install 'SomePackage>=1.0.4'     # minimum version
+          python -m pip install SomePackage            # latest version
+          python -m pip install SomePackage==1.0.4     # specific version
+          python -m pip install 'SomePackage>=1.0.4'     # minimum version
+
+      .. group-tab:: Windows
+
+        .. code-block:: shell
+
+          py -m pip install SomePackage            # latest version
+          py -m pip install SomePackage==1.0.4     # specific version
+          py -m pip install 'SomePackage>=1.0.4'     # minimum version
```
```suggestion
            python -m pip install SomePackage            # latest version
            python -m pip install SomePackage==1.0.4     # specific version
            python -m pip install 'SomePackage>=1.0.4'   # minimum version

      .. group-tab:: Windows

         .. code-block:: shell

            py -m pip install SomePackage            # latest version
            py -m pip install SomePackage==1.0.4     # specific version
            py -m pip install 'SomePackage>=1.0.4'   # minimum version
```

shireenrao commented:
@McSinyx - Sorry about that. I can fix the indent issues. Will do these changes today and re-submit.

shireenrao commented:
I am not sure if what I did was the best way. Accepting all of @McSinyx suggested changes caused a commit for each change. Some changes got outdated so couldnt commit them from Github UI. I am changing them from my local branch. Was this approach ok? Is there any other way of dealing with such a situation?

hugovk commented:
For future reference, it's possible to add multiple suggestions to a batch, and commit in one go. 

* https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

shireenrao commented:
@McSinyx / @uranusjr - One of the checks seem to be stuck. Any thoughts on what I should do here? Should I close the PR and Open it again? Thank you!

McSinyx commented:
@shireenrao, [the branch check by Travis actually passed](https://travis-ci.com/github/pypa/pip/builds/182958585), but GitHub failed to report it.  Since it's not required for merging, you don't need to worry about it.  That being said, I'd be grateful if you can squash all the commits applied from my suggestion (and maybe remove the `Co-Author-By` me, since they're purely whitespace changes so I don't really have any authorship there).

pradyunsg commented:
+1 to squashing the commits.

shireenrao commented:
@McSinyx - I have squashed all the commits based on your suggestion. Please let me know if it looks good.

McSinyx commented on docs/html/reference/pip.rst:
```diff
@@ -221,9 +230,21 @@ Build Options
-included in the command as follows::
+included in the command as follows:
+
+.. tabs::
+
+  .. group-tab:: Unix/macOS
+
+    .. code-block:: shell
+
+        python setup.py <global_options> BUILD COMMAND <build_options>
+
+  .. group-tab:: Windows
+
+    .. code-block:: shell
```
I think we missed this.

McSinyx commented on docs/html/reference/pip_download.rst:
```diff
@@ -56,64 +64,148 @@ Examples
-          --python-version 27 \
-          --implementation cp \
-          SomePackage
+        .. code-block:: shell
```
```suggestion
         .. code-block:: shell
```

McSinyx commented on docs/html/reference/pip_download.rst:
```diff
@@ -56,64 +64,148 @@ Examples
-          SomePackage
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
```suggestion
         .. code-block:: shell
```

McSinyx commented on docs/html/reference/pip_download.rst:
```diff
@@ -56,64 +64,148 @@ Examples
-          SomePackage
+      .. group-tab:: Windows
+
+        .. code-block:: shell
```
Also the ones that follow within this file.

```suggestion
         .. code-block:: shell
```

McSinyx commented on docs/html/reference/pip_list.rst:
```diff
@@ -32,75 +40,182 @@ Examples
-    ::
+    .. tabs::
+
+      .. group-tab:: Unix/macOS
+
+        .. code-block:: console
```
This file is not well indented either.  At this point I suggest to either:

* Let your editor/linter autoindent everything if it has such functionality
* Leave it as is for follow-up with auto reST linter

shireenrao replied:
I will fix these shortly.

McSinyx commented on docs/html/reference/pip_uninstall.rst:
```diff
@@ -30,11 +38,26 @@ Examples
-  ::
+  .. tabs::
+
+      .. group-tab:: Unix/macOS
```
I've just went through the files and it seems that this (till the end of file) is the only one you missed.  Thank you for putting up with my OCD :sweat_smile:  Personally I can't wait to see this merged!

shireenrao replied:
No issues at all. Just submitted this change.

shireenrao commented:
@pradyunsg - Do you want me to squash all the commits in this PR to a single commit?

pradyunsg commented:
It's fine -- Travis CI is acting up.

pradyunsg commented:
At the cost of being a little annoying... I'd like us to defer adding sphinx-tabs to our current docs, since [sphinx-panels](https://sphinx-panels.readthedocs.io) added the same functionality and that one can be stylized to look real nice much more easily!

Example from [Furo's documentation](https://pradyunsg.me/furo/customisation/):

![2020-09-16 18 23 56](https://user-images.githubusercontent.com/3275593/93339943-d2228480-f849-11ea-8545-8b8b3ef017b2.gif)

McSinyx commented:
Personally I'd favor sphinx-tabs though, for the reason mentioned in the docs @pradyunsg linked:

> The key difference is that, whereas sphinx-tabs uses JavaScript to implement this functionality, sphinx-panels only uses CSS. A CSS only solution has the benefit of faster load-times, and working when JS is disabled, although **JS allows sphinx-tabs to implement some extended functionality (like synchronized selections)**.

Regarding the look, I think we can work with upstream to see if there's some option, but I'd consider being able to select the OS just one a must-have feature.

pradyunsg commented:
> I think we can work with upstream

I'm an upstream maintainer/collaborator on both of those packages. :)

McSinyx commented:
Is there any technical hindrance preventing sphinx-tabs to be styled (TBH I'm fine with the way it looks ATM)?  Also the entry point invocation issue has caused quite many issues since I started watching this repo that I think it'd be *really* nice to have this merged as soon as possible, especially when @shireenrao has put *a lot* of effort following up on the patch.

pfmoore commented:
> I'd consider being able to select the OS just one a must-have feature.

I'm not 100% sure what you meant here, but the current version lets you select the OS once per page, but if you go from (for example) "Installing" to "User Guide", you need to re-select your preferred OS. For Unix users not such a big deal (as it's the default) but annoying for Windows.

Two possibilities here:

1. Remember the user's preference (via a cookie).
2. Detect the user's OS via browser sniffing.

Currently, this PR feels like it's "nearly there". The defaulting isn't perfect, and we can argue styling all day (I'm not massively keen on the current choice of style for example, but 🤷 ) If the OS selection/persistence issue can be corrected, I'd like to see that. If it's hard, then I'd like to see a follow-up issue raised saying that it needs fixing, so it doesn't get forgotten.

But let's not make perfect be the enemy of good. I'd be fine with merging this and doing subsequent work (including something like switching the implementation to `sphinx-panels`) in follow-up PRs. I agree that @shireenrao has done a lot of good work here, so it would be nice to see that merged so users can benefit.

McSinyx commented:
Thank you for the correction, I was arguing that design-wise sphinx-panels wouldn't support tab-group equivalence.  I'll file the issue on site-wise persistence just in case it is forgotten a few days after this getting merged.

pradyunsg commented:
Let's merge this in. This PR is an improvement over status quo, and any potential future changes should likely build up on top of this one. :)

shireenrao commented:
I was playing around with the look and had posted some styles earlier [here](https://github.com/pypa/pip/pull/8589#issuecomment-662054242).
The look can be adjusted easily. 

As for the issue with remembering the tabs across pages.. there was an [issue opened](https://github.com/executablebooks/sphinx-tabs/issues/80) a few weeks ago which should bring this feature in.

pradyunsg commented:
The issue you've linked to is blocked on me doing a rewrite of the sphinx-tabs package, and I certainly plan to have better styling by default when I get to it.

It might just end up being that I do a rewrite that fixes all these issues for us. 🤷🏻‍♂️

pradyunsg commented:
Thanks @shireenrao! ^>^

uranusjr commented:
Thanks so much for pushing this through @shireenrao!

shireenrao commented:
Is am glad this worked out.

shireenrao commented:
There seems to be a problem on the [Quickstart](https://pip.pypa.io/en/latest/quickstart/) guide. For some reason the tabs are not rendering on that page. The [PR build](https://pip--8589.org.readthedocs.build/en/8589/quickstart/) version is showing it.

hugovk commented:
Good work!

---

I see the tabs on the Quickstart page.

![image](https://user-images.githubusercontent.com/1324225/94180246-64f49c00-fea6-11ea-8d5d-d832e7094033.png)

You could try clearing your browser cache or using Incognito?

shireenrao commented:
Its working now. What was weird was that the tabs were fine in the other pages except the quickstart page. I had tried it on multiple browsers and devices.

webknjaz commented on docs/html/development/getting-started.rst:
```diff
@@ -38,13 +38,25 @@ To run the pip executable from your source tree during development, install pip
+
+   .. group-tab:: Unix/macOS
+
+      .. code-block:: shell
```
@shireenrao @pradyunsg I learned recently that there's a better option of terminal dialogues.
It should be `shell-session` and all the commands should have prompts. This highlighter knows to distinguish between commands and their outputs then, and highlights $-prompt separately

shireenrao replied:
@webknjaz - we have had that discussion above and the consensus was to not have prompts added. Initially I had `C:\>` for windows and `$` for linux/mac. I then took them all out.

webknjaz replied:
But it's present in many other files: https://github.com/pypa/pip/pull/8589/files#diff-25ae4181a2c6592acbec442969beb23cR131 still uses `console` and prompt, for example

pradyunsg replied:
@webknjaz If you'd like to have an extended discussion on this, please file a new issue.

webknjaz replied:
No, just wanted to point out what I noticed and see if anybody cares :)
